File: CS/WebSite/App_Code/XPOHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using DevExpress.Xpo.Metadata;
using DevExpress.Xpo.DB.Cte;
using System.Xml;

    public class XPOHelper
    {
        UnitOfWork uow;
        IDataLayer dal;
        String strCTEName = "";
        System.Web.UI.Page page;
        public XPServerCollectionSource GetXCS(string sServerName, string sUserName, string sPassword, string sDBName, string sSql, string sColumnXML,
            System.Web.UI.Page oPage, bool p_bHasKeyField)
        {
            XPServerCollectionSource l_xcs = null;
            page = oPage;
            page.Unload += new EventHandler(Page_Unload);
            strCTEName = "CTE_" + DateTime.Now.Hour.ToString() + DateTime.Now.Minute.ToString() + DateTime.Now.Millisecond.ToString();
            //Use your own GetConnectionString method based on the passed parameters
            string connectionString = MSSqlConnectionProviderWithCte.GetConnectionString(sServerName, sDBName);
            XPDictionary dict = new ReflectionDictionary();
            XPClassInfo classInfo = dict.CreateClass(dict.GetClassInfo(typeof(BasePersistentClass)), strCTEName);

            XmlDocument l_xmlDoc = new XmlDocument();
            l_xmlDoc.LoadXml(sColumnXML);
            XmlNodeList l_xnl = l_xmlDoc.SelectNodes("//FLD");
            for (int i = 0; i < l_xnl.Count; i++)
            {
                Type l_type = null;
                switch (l_xnl[i]["DATTYP"].InnerText.ToLower())
                {
                    case "int64":
                        l_type = typeof(Int64);
                        break;
                    case "byte[]":
                        l_type = typeof(Byte[]);
                        break;
                    case "boolean":
                        l_type = typeof(Boolean);
                        break;
                    case "string":
                        l_type = typeof(String);
                        break;
                    case "char[]":
                        l_type = typeof(Char[]);
                        break;
                    case "datetime":
                        l_type = typeof(DateTime);
                        break;
                    case "datetimeoffset":
                        l_type = typeof(DateTimeOffset);
                        break;
                    case "decimal":
                        l_type = typeof(Decimal);
                        break;
                    case "double":
                        l_type = typeof(Double);
                        break;
                    case "int32":
                        l_type = typeof(Int32);
                        break;
                    case "single":
                        l_type = typeof(Single);
                        break;
                    case "int16":
                        l_type = typeof(Int16);
                        break;
                    case "object":
                        l_type = typeof(Object);
                        break;
                    case "timeSpan":
                        l_type = typeof(TimeSpan);
                        break;
                    case "byte":
                        l_type = typeof(Byte);
                        break;
                    case "guid":
                        l_type = typeof(Guid);
                        break;
                    case "xml":
                        l_type = typeof(XmlDocument);
                        break;
                }
                
                if (i == 0)
                    classInfo.CreateMember(l_xnl[i]["COL"].InnerText, l_type, new KeyAttribute(true));
                else
                    classInfo.CreateMember(l_xnl[i]["COL"].InnerText, l_type);
            }

            dal = new SimpleDataLayer(dict, XpoDefault.GetConnectionProvider(connectionString, AutoCreateOption.SchemaAlreadyExists));
            uow = new UnitOfWork(dal);
            string l_strQuery = "";
            l_strQuery = " AS (" + sSql + ")";
            uow.RegisterCte(strCTEName, @l_strQuery);
            try
            {

                l_xcs = new XPServerCollectionSource(uow, classInfo);
                return l_xcs;
            }
            finally
            {

            }
            return l_xcs;
        }

        protected void Page_Unload(object sender, EventArgs e)
        {
            uow.UnregisterCte(strCTEName);
            uow.Dispose();
            dal.Dispose();
            page.Unload -= new EventHandler(Page_Unload);
        }

        [NonPersistent]
        public class BasePersistentClass : XPLiteObject
        {
            public BasePersistentClass(Session session) : base(session) { }
            public BasePersistentClass(Session session, XPClassInfo classInfo) : base(session, classInfo) { }
        }
    }

File: VB/WebSite/App_Code/XPOHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web

Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo.Metadata
Imports DevExpress.Xpo.DB.Cte
Imports System.Xml

	Public Class XPOHelper
		Private uow As UnitOfWork
		Private dal As IDataLayer
		Private strCTEName As String = ""
		Private page As System.Web.UI.Page
		Public Function GetXCS(ByVal sServerName As String, ByVal sUserName As String, ByVal sPassword As String, ByVal sDBName As String, ByVal sSql As String, ByVal sColumnXML As String, ByVal oPage As System.Web.UI.Page, ByVal p_bHasKeyField As Boolean) As XPServerCollectionSource
			Dim l_xcs As XPServerCollectionSource = Nothing
			page = oPage
			AddHandler page.Unload, AddressOf Page_Unload
			strCTEName = "CTE_" & DateTime.Now.Hour.ToString() & DateTime.Now.Minute.ToString() & DateTime.Now.Millisecond.ToString()
			'Use your own GetConnectionString method based on the passed parameters
			Dim connectionString As String = MSSqlConnectionProviderWithCte.GetConnectionString(sServerName, sDBName)
			Dim dict As XPDictionary = New ReflectionDictionary()
			Dim classInfo As XPClassInfo = dict.CreateClass(dict.GetClassInfo(GetType(BasePersistentClass)), strCTEName)

			Dim l_xmlDoc As New XmlDocument()
			l_xmlDoc.LoadXml(sColumnXML)
			Dim l_xnl As XmlNodeList = l_xmlDoc.SelectNodes("//FLD")
			For i As Integer = 0 To l_xnl.Count - 1
				Dim l_type As Type = Nothing
				Select Case l_xnl(i)("DATTYP").InnerText.ToLower()
					Case "int64"
						l_type = GetType(Int64)
					Case "byte[]"
						l_type = GetType(Byte())
					Case "boolean"
						l_type = GetType(Boolean)
					Case "string"
						l_type = GetType(String)
					Case "char[]"
						l_type = GetType(Char())
					Case "datetime"
						l_type = GetType(DateTime)
					Case "datetimeoffset"
						l_type = GetType(DateTimeOffset)
					Case "decimal"
						l_type = GetType(Decimal)
					Case "double"
						l_type = GetType(Double)
					Case "int32"
						l_type = GetType(Int32)
					Case "single"
						l_type = GetType(Single)
					Case "int16"
						l_type = GetType(Int16)
					Case "object"
						l_type = GetType(Object)
					Case "timeSpan"
						l_type = GetType(TimeSpan)
					Case "byte"
						l_type = GetType(Byte)
					Case "guid"
						l_type = GetType(Guid)
					Case "xml"
						l_type = GetType(XmlDocument)
				End Select

				If i = 0 Then
					classInfo.CreateMember(l_xnl(i)("COL").InnerText, l_type, New KeyAttribute(True))
				Else
					classInfo.CreateMember(l_xnl(i)("COL").InnerText, l_type)
				End If
			Next i

			dal = New SimpleDataLayer(dict, XpoDefault.GetConnectionProvider(connectionString, AutoCreateOption.SchemaAlreadyExists))
			uow = New UnitOfWork(dal)
			Dim l_strQuery As String = ""
			l_strQuery = " AS (" & sSql & ")"
			uow.RegisterCte(strCTEName, l_strQuery)
			Try

				l_xcs = New XPServerCollectionSource(uow, classInfo)
				Return l_xcs
			Finally

			End Try
			Return l_xcs
		End Function

		Protected Sub Page_Unload(ByVal sender As Object, ByVal e As EventArgs)
			uow.UnregisterCte(strCTEName)
			uow.Dispose()
			dal.Dispose()
			RemoveHandler page.Unload, AddressOf Page_Unload
		End Sub

		<NonPersistent> _
		Public Class BasePersistentClass
			Inherits XPLiteObject
			Public Sub New(ByVal session As Session)
				MyBase.New(session)
			End Sub
			Public Sub New(ByVal session As Session, ByVal classInfo As XPClassInfo)
				MyBase.New(session, classInfo)
			End Sub
		End Class
	End Class
